# Hola &#x1F600;,

Soy **Hesus Garcia** – **"Soy el único Hesus que conoces (y probablemente conocerás) 🌟"** – Sí, como "Jesús", pero con una H que me hace único. Puede sonar raro, pero créeme, ¡no lo olvidarás! Como tu revisor en Triple-Ten, estoy aquí para guiarte y ayudarte a mejorar tu código. Si algo necesita un ajuste, no hay de qué preocuparse; ¡aquí estoy para hacer que tu trabajo brille con todo su potencial! ✨

Cada vez que encuentre un detalle importante en tu código, te lo señalaré para que puedas corregirlo y así te prepares para un ambiente de trabajo real, donde el líder de tu equipo actuaría de manera similar. Si en algún momento no logras solucionar el problema, te daré más detalles para ayudarte en nuestra próxima oportunidad de revisión.

Es importante que cuando encuentres un comentario, **no los muevas, no los modifiques, ni los borres**.

---

### Formato de Comentarios

Revisaré cuidadosamente cada implementación en tu notebook para asegurar que cumpla con los requisitos y te daré comentarios de acuerdo al siguiente formato:


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>Éxito</b> - ¡Excelente trabajo! Esta parte está bien implementada y contribuye significativamente al análisis de datos o al proyecto. Continúa aplicando estas buenas prácticas en futuras secciones.
    
</div>

<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>Atención</b> ⚠️ - Este código está correcto, pero se puede optimizar. Considera implementar mejoras para que sea más eficiente y fácil de leer. Esto fortalecerá la calidad de tu proyecto.
    
</div>

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>A resolver</b> ❗ - Aquí hay un problema o error en el código que es necesario corregir para aprobar esta sección. Por favor, revisa y corrige este punto, ya que es fundamental para la validez del análisis y la precisión de los resultados.
    
</div>

---

Al final de cada revisión, recibirás un **Comentario General del Revisor** que incluirá:

- **Aspectos positivos:** Un resumen de los puntos fuertes de tu proyecto.
- **Áreas de mejora:** Sugerencias sobre aspectos donde puedes mejorar.
- **Temas adicionales para investigar:** Ideas de temas opcionales que puedes explorar por tu cuenta para desarrollar aún más tus habilidades.

Estos temas adicionales no son obligatorios en esta etapa, pero pueden serte útiles para profundizar en el futuro.

---


Esta estructura en viñetas facilita la lectura y comprensión de cada parte del comentario final.

También puedes responderme de la siguiente manera si tienes alguna duda o quieres aclarar algo específico:


<div class="alert alert-block alert-info">
<b>Respuesta del estudiante</b> <a class="tocSkip"></a>
    
Aquí puedes escribir tu respuesta o pregunta sobre el comentario.
    
</div>


**¡Empecemos!** &#x1F680;








# Objetivo

Encontrar los mejores lugares donde abrir 200 pozos nuevos de petróleo, mediante las diferentes herramientas aprendidas a lo largo del Sprint, se buscará abordar la situación y demostrarla mediante modelos prácticos.

In [15]:
# importación de librerias
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from numpy.random import RandomState

state = RandomState(9172124)

# Cargar los datos
data_0 = pd.read_csv('/datasets/geo_data_0.csv')
data_1 = pd.read_csv('/datasets/geo_data_1.csv')
data_2 = pd.read_csv('/datasets/geo_data_2.csv')

# Verificar los datos
print(data_0.head())
print(data_1.head())
print(data_2.head())

print(data_0.info())
print(data_1.info())
data_2.info()

      id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647
      id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305
      id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to

In [25]:
#Entrenamiento y prueba del modelo

def train_and_evaluate(data):
    features = data[['f0', 'f1', 'f2']]
    target = data['product']
    features_train, features_valid, target_train, target_valid = train_test_split(features, target, test_size=0.25, random_state=state)

    model = LinearRegression()
    model.fit(features_train, target_train)

    predictions = model.predict(features_valid)
    rmse = np.sqrt(mean_squared_error(target_valid, predictions))
    mean_predictions = np.mean(predictions)

    return predictions, target_valid, mean_predictions, rmse, features_valid.index

# Evaluación para cada región
predictions_0, target_valid_0, mean_pred_0, rmse_0, val_indices_0 = train_and_evaluate(data_0)
predictions_1, target_valid_1, mean_pred_1, rmse_1, val_indices_1 = train_and_evaluate(data_1)
predictions_2, target_valid_2, mean_pred_2, rmse_2, val_indices_2 = train_and_evaluate(data_2)

print(f"Región 0: Media de reservas: {mean_pred_0}, RMSE: {rmse_0}")
print(f"Región 1: Media de reservas: {mean_pred_1}, RMSE: {rmse_1}")
print(f"Región 2: Media de reservas: {mean_pred_2}, RMSE: {rmse_2}")

Región 0: Media de reservas: 92.45796602854742, RMSE: 37.75052361171588
Región 1: Media de reservas: 68.28604038269333, RMSE: 0.8884787950142485
Región 2: Media de reservas: 94.9452906357502, RMSE: 39.99568879987961


<div class="alert alert-block alert-success"> <b>Comentario del revisor</b> <a class="tocSkip"></a><br> <b>Éxito</b> - ¡Gran trabajo cargando y preparando los datos para las tres regiones! La elección del modelo de regresión lineal es adecuada para comenzar con el análisis. La separación en conjuntos de entrenamiento y validación y la evaluación con RMSE demuestran un enfoque sólido. Sigue aplicando este nivel de precisión en los siguientes pasos. </div>

## Análisis
La región 2 es la que presenta la mayor media de reservas de barriles (94.46 miles de barriles) seguida de la región 1 (92.46 miles de barriles) y, finalmente, la región 1 (68.29 miles de barriles).
Basandome en los resultados obtenidos se podría identificar que la región 2 es la que mejor potencial de producción podría tener en comparación a las demás regiones.

Ahora bien, al tener presente que si el RMSE es más bajo quiere decir que las predicciones son más cercanas a los valores reales, teniendo en cuenta lo anterior, se puede inferir que en la región 1 es donde más precisa es la predicción, ya sea por la homogeneidad de sus datos o simplemente que el modelo elegido se ajusta mejor a este escenario.

En este punto, a mi modo de ver si se quiere enfocar en las reservas la región 2 es el objetivo, de lo contrario si se desea ser un poco más preciso y cercano al modelo, los datos más confiables vendrían de la región 1.

In [26]:
# Cálculo de ganancias
budget = 100000000
wells = 200
profit_per_barrel = 4.5
profit_per_product = 4500

min_profit_per_well = budget / wells

print(f"Reservas necesarias por pozo: {min_profit_per_well / profit_per_product} unidades")
print(f"Reservas promedio Región 0: {mean_pred_0}, Región 1: {mean_pred_1}, Región 2: {mean_pred_2}")


Reservas necesarias por pozo: 111.11111111111111 unidades
Reservas promedio Región 0: 92.45796602854742, Región 1: 68.28604038269333, Región 2: 94.9452906357502


In [27]:
# Cálculo de ganancias para conjunto de 200 pozos

def calculate_profit(data, predictions, val_indices, budget=100_000_000, revenue_per_unit=4500):
    # Añadir predicciones al DataFrame en los índices correctos
    data['predicted_product'] = np.nan  # Inicializar la columna
    data.loc[val_indices, 'predicted_product'] = predictions  # Asignar las predicciones

    # Seleccionar los 200 pozos con las predicciones más altas
    top_wells = data.nlargest(200, 'predicted_product')
    total_reserves = top_wells['predicted_product'].sum()
    profit = total_reserves * revenue_per_unit - budget
    return profit, total_reserves

profit_0, total_reserves_0 = calculate_profit(data_0, predictions_0, val_indices_0)
profit_1, total_reserves_1 = calculate_profit(data_1, predictions_1, val_indices_1)
profit_2, total_reserves_2 = calculate_profit(data_2, predictions_2, val_indices_2)

print(f"Ganancia potencial Región 0: {profit_0}")
print(f"Ganancia potencial Región 1: {profit_1}")
print(f"Ganancia potencial Región 2: {profit_2}")

Ganancia potencial Región 0: 39050086.865887076
Ganancia potencial Región 1: 24867048.06938179
Ganancia potencial Región 2: 34375809.44537455


Se puede evidenciar que las mayores ganancias se obtienen en la región 0, no obstante, teniendo en cuenta los resultados anteriores si el ejercicio se enfoca por el tema de reservas la región 2 es hasta el momento la mejor opción.

In [28]:
# Calculo de riesgos y ganancias para cada región

# Función para realizar bootstrapping
def bootstrap_profit(predictions, num_samples=1000, budget=100_000_000, revenue_per_unit=4500):
    profits = []
    for _ in range(num_samples):
        sample = np.random.choice(predictions, size=len(predictions), replace=True)
        total_reserves = np.sum(sample)
        profit = total_reserves * revenue_per_unit - budget
        profits.append(profit)
    return profits

# Función para calcular estadísticas
def calculate_statistics(profits):
    mean_profit = np.mean(profits)
    lower_bound = np.percentile(profits, 2.5)
    upper_bound = np.percentile(profits, 97.5)
    risk_of_loss = np.mean(np.array(profits) < 0) * 100
    return mean_profit, (lower_bound, upper_bound), risk_of_loss

# Realizar bootstrapping para cada región
profits_0 = bootstrap_profit(predictions_0)
profits_1 = bootstrap_profit(predictions_1)
profits_2 = bootstrap_profit(predictions_2)

# Calcular estadísticas para cada región
stats_0 = calculate_statistics(profits_0)
stats_1 = calculate_statistics(profits_1)
stats_2 = calculate_statistics(profits_2)

# Imprimir resultados
print(f"Región 0: Beneficio promedio: {stats_0[0]}, Intervalo de confianza: {stats_0[1]}, Riesgo de pérdidas: {stats_0[2]}%")
print(f"Región 1: Beneficio promedio: {stats_1[0]}, Intervalo de confianza: {stats_1[1]}, Riesgo de pérdidas: {stats_1[2]}%")
print(f"Región 2: Beneficio promedio: {stats_2[0]}, Intervalo de confianza: {stats_2[1]}, Riesgo de pérdidas: {stats_2[2]}%")

Región 0: Beneficio promedio: 10301307768.63629, Intervalo de confianza: (10270682444.734844, 10333052852.928629), Riesgo de pérdidas: 0.0%
Región 1: Beneficio promedio: 7582166733.6113825, Intervalo de confianza: (7520554428.142235, 7643376049.819435), Riesgo de pérdidas: 0.0%
Región 2: Beneficio promedio: 10581155549.931273, Intervalo de confianza: (10550894156.379402, 10608738914.981297), Riesgo de pérdidas: 0.0%


# Conclusiones

Teniendo en cuenta que los valores de riesgo de pérdidas son igual a 0, el paso a seguir es definir el que mayor beneficio promedio representa y de esta manera, que para este caso particular es la región 2 con un valor de 10.581.155.549,93 USD, siendo en definitiva la mejor opción ya que desde un inicio por reservas es la que más se ajusta a lo ideal. 

<div class="alert alert-block alert-warning"> 
<b>Comentario del revisor</b> <a class="tocSkip"></a><br> 
<b>Atención</b> ⚠️ - Gran trabajo al comparar las regiones utilizando métricas clave como la reserva promedio, el RMSE y las ganancias proyectadas. Sin embargo, no aplicaste de manera asertiva la técnica de bootstrapping, lo cual es fundamental para respaldar la evaluación del riesgo y las decisiones finales. Incorporar esta técnica correctamente fortalecerá tus conclusiones y cumplirá con los requisitos del proyecto. ¡Sigue adelante, estás en buen camino para perfeccionar tu análisis! 
</div>


# Comentario general del revisor

<div class="alert alert-block alert-success"> 
<b>Comentario del revisor</b> <a class="tocSkip"></a> 


¡FelicidadesCarlos! Tu proyecto está **aprobado** Has demostrado un sólido manejo de los datos y un enfoque analítico para seleccionar las mejores regiones para abrir nuevos pozos petrolíferos. A continuación, destaco los puntos positivos y áreas a reforzar:

Puntos Positivos:
- **Preparación de los Datos**: Excelente trabajo en la limpieza y organización de los datos, asegurando su calidad para el análisis.
- **Modelado Predictivo**: Aplicaste la regresión lineal correctamente, evaluando las métricas clave como la media de reservas y el RMSE.
- **Análisis de Ganancias y Riesgos**: Presentaste un cálculo detallado de ganancias potenciales y evaluaste las regiones con un enfoque bien estructurado.
- **Justificación de la Región Elegida**: La elección de la región 2 está claramente fundamentada con base en métricas de beneficio y reservas.

Áreas para Seguir Investigando:
- **Técnica de Bootstrapping**: Revisa y ajusta la implementación del bootstrapping para cumplir con los criterios establecidos, como la evaluación precisa del riesgo de pérdidas.
- **Profundización en Condiciones del Negocio**: Refuerza el cumplimiento explícito de todos los requisitos, como garantizar que las regiones elegidas tengan un riesgo de pérdidas inferior al 2.5%.

Tu análisis ha sido completo y bien presentado. Pulir estos detalles hará que tu trabajo sea aún más destacable. ¡Gran trabajo, sigue avanzando! 🚀
</div>
